notes/plutus-core/cek-budgeting-profiling/CostingDetails.md

```diff
@@ -0,0 +1,240 @@
+    spendBudget key budget = do
+        modifying exBudgetStateTally
+                (<> (ExTally (singleton key budget)))
+        newBudget <- exBudgetStateBudget <%= (<> budget)
+        mode <- asks cekEnvBudgetMode
+        case mode of
+            Counting -> pure ()
+            Restricting resb ->
+                when (exceedsBudget resb newBudget) $
+                    throwingWithCause _EvaluationError
+                        (UserEvaluationError $ CekOutOfExError resb newBudget)
+                        Nothing  -- No value available for error
+```
+
+The machine runs in a monad containing a `CekBudgetState` object recording
+costs.  Note that `spendBudget` updates both `_exBudgetStateTally` and
+`_exBudgetStateBudget`: the `modifying` line updates the accumulated
+per-operation costs with the costs of the current operation, and the line
+`newBudget <- exBudgetStateBudget <%= (<> budget)` adds the cost of the current
+operation to the total consumed so far (and returns the updated result for
+further use).  Both fields are updated irrespective of
+whether we're in `Counting` mode or `Restricting` mode, which may explain why
+the figures in [CekProfiling.md](./CekProfiling.md) don't show much difference
+between execution times for the two modes.  To attempt to remedy this, it seems
+sensible to refactor `spendBudget` so that it's only updating the field relevant
+to the current mode:
+
+```
+    spendBudget key budget = do
+        mode <- asks cekEnvBudgetMode
+        case mode of
+            Counting -> modifying exBudgetStateTally (<> (ExTally (singleton key budget)))
+            Restricting resb -> do
+                newBudget <- exBudgetStateBudget <%= (<> budget)
+                when (exceedsBudget resb newBudget) $
+                    throwingWithCause _EvaluationError
+                        (UserEvaluationError $ CekOutOfExError resb newBudget)
+                        Nothing
+```
+
+This does reduce execution times noticeably in `Restricting` mode (see table
+later), but for some reason it reintroduces a memory leak in `Counting` mode:
+many of the `nofib` examples fail to complete because they consume large amounts
+of memory very quickly.  Changing the code to 
+
+```
+    spendBudget key budget = do
+        newBudget <- exBudgetStateBudget <%= (<> budget)
+        mode <- asks cekEnvBudgetMode
+        case mode of
+            Counting -> modifying exBudgetStateTally (<> (ExTally (singleton key budget)))
+            Restricting resb -> 
+                when (exceedsBudget resb newBudget) $
+                    throwingWithCause _EvaluationError
+                        (UserEvaluationError $ CekOutOfExError resb newBudget)
+                        Nothing
+```
+
+eliminates the leak again, but it's not clear to me why.  Possibly `modifying`
+is accumulating costs lazily and updating `_exBudgetStateBudget` prevents this
```

Comment:
As per a slack discussion, yeah, `modifying` calls the lazy `modify` instead of the strict `modify'`.

Reply:
:+1: I've added a note about that.